Enable TS strict mode for React Aria Components

devongovett opened the pull request:
Builds on #4212.

This converts the code and examples for React Aria Components to TS strict mode. It also improves the types for:

* Date picker components, by making stately hooks generic. Also adds `null` to the `value` prop types. Part of https://github.com/adobe/react-spectrum/issues/3187.
* Adds null to Collection and Node types.
* Omits `children` as a required prop for `useTabList`.
* Adds `isTextDropItem`, `isFileDropItem`, and `isDirectoryDropItem` functions which act as TS [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).

snowystinger commented on packages/react-aria-components/src/Collection.tsx:
```diff
@@ -484,6 +487,7 @@ export class Document<T, C extends BaseCollection<T>> extends BaseNode<T> {
   private subscriptions: Set<() => void> = new Set();
 
   constructor(collection: C) {
+    // @ts-ignore
```
this one stumps me, why is this ok?

devongovett replied:
It is a circular dependency... the `ownerDocument` needs to be itself, but `this` cannot be accessed until after calling `super`. It gets assigned in the initializer above, and the super constructor doesn't actually use it other than assigning so figured ignoring was ok.

LFDanLu reviewed (approved):
LGTM

devongovett commented:
Merged in main, and updated the drag and drop examples and code to TS strict as well. In order to make this nicer, I also added `isTextDropItem`, `isFileDropItem`, and `isDirectoryDropItem` functions which act as TS [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) as suggested in https://github.com/adobe/react-spectrum/discussions/3765. This makes it easier to filter drop items by kind, while enabling TypeScript to narrow the type from `DropItem` to `TextDropItem` etc. for use in subsequent array mapping.

ktabors commented on packages/react-aria-components/docs/ListBox.mdx:
````diff
@@ -842,8 +842,13 @@ function DraggableListBox() {
 Dropping on the ListBox as a whole can be enabled using the `onRootDrop` event. When a valid drag hovers over the ListBox, it receives the `isDropTarget` state and can be styled using the `[data-drop-target]` CSS selector.
 
 ```tsx example
+interface Item {
````
I feel like this should be a type we export since it's used a couple times, but it's just for examples to show what to do.
